### PR TITLE
eth/downloader: track peer drops and deassign state sync tasks

### PR DESCRIPTION
**Pull request by karalabe**
Previously when the state sync was part of the downloader queue, peer drops resulted in successful cancellation of pending state sync tasks too. However with the state sync rewrite, the state download tasks were not cancelled on peer drop, rather relying on timeouts to detect such scenarios.

Apparently there are two issues with this approach:

 * If a peer is dropped for not responding to an arbitrary non-state-download request (or simply disconnects remotely), the state task will not be rescheduled until it times out. This might be up to 2 minutes for slow connections. That can have a seriously adverse effect on performance.
 * If the peer reconnects in between, it will be assigned a new download task, and the old one scheduled for discarding (emulating a timeout on it). This can quite easily trigger a new disconnect since the downloader things the peer is stalling, not disconnected and freshly reconnected. This causes peer churn.

The PR fixes the above issues by tracking peer departure events too. If a peer is departed, any associated state download task is marked as "failed due to peer drop", and rescheduled for someone else to download. Similarly if a peer is double-assigned (should not happen any more, but just in case), the old task is not considered timed out as until now, rather it's emulated as a peer drop, ensuring the peer doesn't get disconnected due to some error/event race.

### CODE REVIEW

File: eth/downloader/peer.go [hunk 1]
```diff
@@ -361,10 +362,16 @@ func newPeerSet() *peerSet {
 	}
 }
 
+// SubscribeNewPeers subscribes to peer arrival events.
 func (ps *peerSet) SubscribeNewPeers(ch chan<- *peerConnection) event.Subscription {
 	return ps.newPeerFeed.Subscribe(ch)
 }
 
+// SubscribeOldPeers subscribes to peer departure events.
+func (ps *peerSet) SubscribeOldPeers(ch chan<- *peerConnection) event.Subscription {
```
Comment (fjl) [inline eth/downloader/peer.go]: `SubscribePeerDrop`

Comment (karalabe) [reply eth/downloader/peer.go]: Fixed, PTAL

**Comment by JulyKe** [comment]
Every time the peer reconnects, it will construct stateSync data structure freely. Even the download task was added to the retry queue (oldStateSync.tasks[hash]=task), it will not affect the new peer connection. I know your logic in this PR but it seems not a problem in the real experiments.

**Comment by JulyKe** [comment]
Every time the peer reconnects, it will construct stateSync data structure freely. Even the download task was added to the retry queue (oldStateSync.tasks[hash]=task), it will not affect the new peer connection. I know your logic in this PR but it seems not a problem in the real experiments.

**Comment by JulyKe** [comment]
Every time the peer reconnects, it will construct stateSync data structure freely. Even the download task was added to the retry queue (oldStateSync.tasks[hash]=task), it will not affect the new peer connection. I know your logic in this PR but it seems not a problem in the real experiments.

**Comment by JulyKe** [comment]
Based on my experiments, if you want to reproduce this bug, it's not only relate to state sync, but also depends on the BlockHeaderMsg.

You should get partial block headers  (e.g., 23 out of 37) before drop the peer, then reconnect, it will construct the state trie based on the previous received block header messages, which is not the latest block header, then the state trie might be different from the state trie of block 37. Then even reconnecting I get the state trie, but it is missing something, which lead to the td less than the expected, then returns errStallingPeer exception and drop the peer again. and then repeat.

**Pull request merged**